### PR TITLE
NFT auto detect metrics

### PR DESCRIPTION
**Pull request by adonesky1**
Addresses subtasks on https://github.com/MetaMask/metamask-extension/issues/13477

**BREAKING** - add method `trackMetaMetricsEvent` passed to `CollectiblesController` in options object to and call it to broadcast `Token Added` event when a `collectible`/`nft` is added to state.

### CODE REVIEW

**Review by brad-decker — CHANGES_REQUESTED** [review]
[REQ] I think we should, instead of passing the metrics handler in (which will only apply to the extension API at present) we should instead add events/hooks/whatever into the places where we expect metrics to be fired. 

Reasoning:
1. The MetaMetrics controller is slated to be lifted to this repo at some point in the future, when that happens the API will likely adjust slightly.
2. If we think about those points in the flow we might wanna track metrics, etc then we can use the messaging system of this repo to have the MetaMetrics controller listen for those events and fire metrics accordingly, keeping most metrics related code in the metrics controller.
3. Until 1 + 2 it gives some un-opinionated options for mobile to hook into to do metrics until they are aligned with the metrics api.

**Comment by adonesky1** [comment]
> Until 1 + 2 it gives some un-opinionated options for mobile to hook into to do metrics until they are aligned with the metrics api.

A less opinionated hook/event would still need to take as an arg object containing all the `sensitiveProperties` we're passing here:

```
 sensitiveProperties: {
          token_contract_address: address,
          token_symbol: collectibleContract.symbol,
          tokenId: tokenId.toString(),
          asset_type: ASSET_TYPES.COLLECTIBLE,
          token_standard: collectibleMetadata.standard,
          source: detection ? 'detected' : 'custom',
        },
```
It looks like currently mobile uses an event tracking method similar but not congruent to the method signature here:
```
/**
 * This takes params with the following structure:
 * { foo : 'this is not anonymous', bar: {value: 'this is anonymous', anonymous: true} }
 * @param {Object} eventName
 * @param {Object} params
 */
export const trackEventV2 = (eventName, params) => {
 ...
 ```
 
 So I suppose I understand your desired approach to be we pass a method called something like `trackNFTAddedEvent` that takes an object containing the properties we want to track and then pass it to CollectiblesController like this:
 
 ```
  trackMetaMetricsEvent: (params) => this.metaMetricsController.trackEvent({ 
      event: 'Token Added',
      category: 'Wallet',
      sensitiveProperties: params
      }),
 ```
 
 Does that look right to you? @brad-decker

**Comment by brad-decker** [comment]
@adonesky1 I was thinking we abstract the concept of meta metrics entirely, instead we emit a general `token added` event of some sort (or provide a listener for such events, i'm not sure what the correct approach is right now but I do know the messaging system is in place in this repo), then the metametrics controller (on the extension now, later this repo), can listen to that event and fire the tracking. This moves all things metrics into the metametrics controller instead of scattered through the other controllers and will ultimately lead to a single team owning a single implementation of tracking across products.

**Comment by adonesky1** [comment]
> i'm not sure what the correct approach is right now but I do know the messaging system is in place in this repo)

I believe what you are suggesting would involve migrating either one or both of `CollectibleDetectionController` and `MetaMetricsController` to the BaseControllerV2 standard which feels like a big jump in scope...? @Gudahtt Is that right?

**Comment by brad-decker** [comment]
@adonesky1 then we can use the listener pattern? Pass in a method for onTokenAdded? I wasnt suggesting we do this all now, rather we do the MVP approach that isnt passing a function signature that only the extension uses into a controller. The future migration path is cleaner if we do this.

**Comment by adonesky1** [comment]
> @adonesky1 then we can use the listener pattern? Pass in a method for onTokenAdded?

Yes sounds good to me. Like this work for you?

**metamask-controller.js**
```
this.collectibleDetectionController = new CollectibleDetectionController(
{
...
 onTokenAdded: (params) => this.metaMetricsController.trackEvent({ 
     event: 'Token Added',
     category: 'Wallet',
     sensitiveProperties: params
     }),
}
)
```

**Comment by brad-decker** [comment]
@adonesky1  yep! Does this seem like a decent approach to you?

**Comment by adonesky1** [comment]
> Does this seem like a decent approach to you?

Yes. I agree that we should leverage the messaging system in the way you describe soon, but for now I think this is the best approach.

**Comment by adonesky1** [comment]
@brad-decker updated [here](https://github.com/MetaMask/controllers/pull/814/commits/c5c9b93d4fceffdea67b60a12cd6495dd1edc310) and [here](https://github.com/MetaMask/controllers/pull/814/commits/2c6531a4a3ae74f47d0671229af00c1e624f6a05)

File: src/assets/CollectiblesController.ts [hunk 1]
```diff
@@ -803,6 +816,15 @@ export class CollectiblesController extends BaseController<
 
   private getERC1155TokenURI: AssetsContractController['getERC1155TokenURI'];
 
+  private onCollectibleAdded: (sensitiveProperties: {
+    token_contract_address: string;
+    token_symbol?: string;
```
Comment (mcmire) [inline src/assets/CollectiblesController.ts:821]: [nit] It doesn't seem like `token_symbol` is ever left out of the object, so would a more appropriate type be `token_symbol: string | undefined`? This is especially relevant now that TypeScript 4.4 has the ability to treat optional properties and undefined properties as separate concepts.

File: src/assets/CollectiblesController.ts [hunk 2]
```diff
@@ -815,6 +837,7 @@ export class CollectiblesController extends BaseController<
    * @param options.getERC721OwnerOf - Get the owner of a ERC-721 collectible.
    * @param options.getERC1155BalanceOf - Gets balance of a ERC-1155 collectible.
    * @param options.getERC1155TokenURI - Gets the URI of the ERC1155 token at the given address, with the given ID.
+   * @param options.onCollectibleAdded - Event hook currently used to emit a track event for the collectible added.
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply src/assets/CollectiblesController.ts:840]: Per my comment below, happy to do this, though I'm inclined to include the comment that this callback is currently used just for tracking purposes just to give anyone reading through it that context.

File: src/assets/CollectiblesController.test.ts [hunk 3]
```diff
@@ -232,6 +235,50 @@ describe('CollectiblesController', () => {
       });
     });
 
+    it('should call track events method correctly when collectible is manually added', async () => {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply src/assets/CollectiblesController.test.ts]: Updated here: https://github.com/MetaMask/controllers/pull/814/commits/e08684a0e372eaca56273ac65b586cdaf68c2e85

File: src/assets/CollectiblesController.test.ts [hunk 4]
```diff
@@ -232,6 +235,50 @@ describe('CollectiblesController', () => {
...
+        asset_type: ASSET_TYPES.COLLECTIBLE,
+        source: 'custom',
+        tokenId: '1',
+        token_contract_address: '0x01',
+        token_standard: 'ERC1155',
+        token_symbol: 'FOO',
+      });
+    });
+
+    it('should call track events method correctly when collectible is added via detection', async () => {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply src/assets/CollectiblesController.test.ts]: Updated here: https://github.com/MetaMask/controllers/pull/814/commits/e08684a0e372eaca56273ac65b586cdaf68c2e85

File: src/assets/CollectiblesController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adonesky1) [reply src/assets/CollectiblesController.ts]: Updated here: https://github.com/MetaMask/controllers/pull/814/commits/784c27e340d981c27ea094c866604303dd5e55c8

File: src/assets/CollectiblesController.ts [hunk 5]
```diff
@@ -841,6 +865,14 @@ export class CollectiblesController extends BaseController<
       getERC721OwnerOf: AssetsContractController['getERC721OwnerOf'];
       getERC1155BalanceOf: AssetsContractController['getERC1155BalanceOf'];
       getERC1155TokenURI: AssetsContractController['getERC1155TokenURI'];
+      onCollectibleAdded: (sensitiveProperties: {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply src/assets/CollectiblesController.ts]: @brad-decker @mcmire updated here to camelCase / purpose agnostic form: https://github.com/MetaMask/controllers/pull/814/commits/60387a0816b5db7413ac47569a000ebdf1b6bfad

File: src/assets/CollectiblesController.ts [hunk 6]
```diff
@@ -815,6 +834,8 @@ export class CollectiblesController extends BaseController<
    * @param options.getERC721OwnerOf - Get the owner of a ERC-721 collectible.
    * @param options.getERC1155BalanceOf - Gets balance of a ERC-1155 collectible.
    * @param options.getERC1155TokenURI - Gets the URI of the ERC1155 token at the given address, with the given ID.
+   * @param options.onCollectibleAdded - Callback that is called when a collectible is added. Currently used pass data
```
Comment (digiwand) [inline src/assets/CollectiblesController.ts:837]: What if we updated the name `onCollectibleAdded` to something like `getDataOnCollectibleAdded` or `onCollectibleAddedCallback`?

Comment (adonesky1) [reply src/assets/CollectiblesController.ts:837]: Open to this.

Comment (digiwand) [reply src/assets/CollectiblesController.ts:837]: erm...please ignore my last comment. After looking through the code more, I realize that the prepending of `on...` is a common pattern for our controller listener callbacks. So `onCollectibleAddedCallback` might be redundant, and starting the name with `get` might throw people off. 

I'm pondering ways to indicate that this method returns a value while also being a callback. Maybe `onCollectibleAddedGetData` or `onCollectibleAddedReturnData`?

This is completely optional.

**Review by digiwand — APPROVED** [review]
This looks good to me! Thank you for taking initiative on this 🙏

**Review by brad-decker — APPROVED** [review]
Sorry for all the change reqs, but i'm happy with how this turned out. Great work :)

**Pull request merged**